lib/rex/post/meterpreter/extensions/stdapi/railgun/library.rb

```diff
@@ -315,6 +316,10 @@ def build_response(packet, function, layouts, arch)
         return_hash['return'] = (rec_return_value != 0)
       when 'VOID'
         return_hash['return'] = nil
+      when 'PCHAR'
+        return_hash['return'] = rec_return_value == 0 ? nil : read_string(rec_return_value)
+      when 'PWCHAR'
+        return_hash['return'] = rec_return_value == 0 ? nil : read_wstring(rec_return_value)
```

Comment:
I got around to testing this. `read_string` and `read_wstring` are not available from this context. I think we need to update `#process_function_call` to forward the `client` (meterpreter instance), object down into here. That will allow us to call `client.util.read_string` and `client.util.read_wstring`.

You can test this by going to Pry from a Meterpreter session and running `railgun.wldap32.ldap_err2string(81)`. It should look like this:

```
meterpreter > pry
[*] Starting Pry shell...
[*] You are in the "client" (session) object

[1] pry(#<Msf::Sessions::Meterpreter_Python_Python>)> railgun.wldap32.ldap_err2string(81)
=> {"GetLastError"=>0, "ErrorMessage"=>"The operation completed successfully.", "return"=>"Server Down"}
[2] pry(#<Msf::Sessions::Meterpreter_Python_Python>)>
```

That shows that the string "Server Down" was the return value.

`#build_response` needs to be updated to take the client object so these two call sites need to be updated to pass it:

* [lib/rex/post/meterpreter/extensions/stdapi/railgun/multicall.rb:93](https://github.com/rapid7/metasploit-framework/blob/e26ab2c202fe6bc6b90e383bee6e20233966c6e5/lib/rex/post/meterpreter/extensions/stdapi/railgun/multicall.rb#L93)
* [lib/rex/post/meterpreter/extensions/stdapi/railgun/library.rb:382](https://github.com/rapid7/metasploit-framework/blob/e26ab2c202fe6bc6b90e383bee6e20233966c6e5/lib/rex/post/meterpreter/extensions/stdapi/railgun/library.rb#L382)

Reply:
See https://github.com/gwillcox-r7/metasploit-framework/pull/3 for these changes.

Reply:
Landed now.